Cubical/Structures/Group/Base.agda

```diff
@@ -0,0 +1,152 @@
+{-# OPTIONS --cubical --no-import-sorts --safe #-}
+module Cubical.Structures.Group.Base where
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Equiv.HalfAdjoint
+open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Univalence
+open import Cubical.Foundations.Transport
+open import Cubical.Foundations.SIP renaming (SNS-PathP to SNS)
+
+open import Cubical.Data.Sigma
+
+open import Cubical.Structures.Macro
+open import Cubical.Structures.NAryOp
+open import Cubical.Structures.Pointed
+open import Cubical.Structures.Semigroup hiding (⟨_⟩)
+open import Cubical.Structures.Monoid hiding (⟨_⟩)
+
+open Iso
+
+private
+  variable
+    ℓ ℓ' : Level
+
+record IsGroup {G : Type ℓ}
+               (0g : G) (_+_ : G → G → G) (-_ : G → G) : Type ℓ where
+
+  constructor isgroup
+
+  field
+    isMonoid  : IsMonoid 0g _+_
+    inverse   : (x : G) → (x + (- x) ≡ 0g) × ((- x) + x ≡ 0g)
+
+  open IsMonoid isMonoid public
+
+  infixl 6 _-_
+
+  _-_ : G → G → G
+  x - y = x + (- y)
+
+  invl : (x : G) → (- x) + x ≡ 0g
+  invl x = inverse x .snd
+
+  invr : (x : G) → x + (- x) ≡ 0g
+  invr x = inverse x .fst
+
+  -- uniqueness of inverse?
+
+record Group : Type (ℓ-suc ℓ) where
+
+  constructor group
+
+  field
+    Carrier : Type ℓ
+    0g      : Carrier
+    _+_     : Carrier → Carrier → Carrier
+    -_      : Carrier → Carrier
+    isGroup : IsGroup 0g _+_ -_
+
+  infix  8 -_
+  infixr 7 _+_
+
+  open IsGroup isGroup public
```

Comment:
Why is this here? I think it should be part of a separate theory module later in the file

Reply:
While I agree it would make more structural sense to put this elsewhere, I have found that often in proof related to groups, I always "open" the lemmas about a specific group, as well as the group notations themselves.
So I find it better to be able to access everything with just
`open Group G` or `module G = Group G`
instead of
```
open Group G
open GroupLemmas G
```
every single time.
I noticed that the HoTT-agda library does the same thing (putting the lemmas in the record definition), see for example https://github.com/HoTT/HoTT-Agda/blob/master/core/lib/types/Group.agda#L32

Should I still move this further down in the file ?

Reply:
I personally prefer to keep the theory separate from the definition just to get better control of what is in scope. I also think the file reads easier this way. For Data/HITs we have Base/Properties files for everything and I think the long term goal should be to organize the algebraic structures like this as well. This could be done in a future PR where we also move them to the Algebra package. Anyone else has an opinion about this? I we plan to do this then it will be a little bit easier if the definition and theory is separated clearly in the file.

Reply:
Another argument against this (at least in the case for groups), is the fact that there may be different definitions, and different ways to "make" a group (for example with makeGroup-left and makeGroup-right), and it is not obvious in this case what should be an axiom and what should be a property.
For example, If I'm doing a proof about multiples groups (say G and H), then I cannot simply write
```
open Group G
open GroupLemmas G
open Group H
open GroupLemmas H
```
because of conflicting names.
So I should instead do something like this :
```
module G = Group G
module Glemmas = GroupLemmas G
module H = Group H
module Hlemmas = GroupLemmas H
```
But then it isn't obvious at all that I should use `G.invr` and `Glemmas.invId` and not the opposite. I think it just adds confusion to the end user experience.

Or maybe we should add all the notations and fields of the record type `Group` to the `GroupLemmas` module.
But I think that too would reduce readability of the code. I think a comment explaining the choice to keep those properties there would make it all clearer and more practical to everyone.

Reply:
> Another argument against this (at least in the case for groups), is the fact that there may be different definitions, and different ways to "make" a group (for example with makeGroup-left and makeGroup-right), and it is not obvious in this case what should be an axiom and what should be a property.
> For example, If I'm doing a proof about multiples groups (say G and H), then I cannot simply write
> 
> ```
> open Group G
> open GroupLemmas G
> open Group H
> open GroupLemmas H
> ```
> 
> because of conflicting names.
> So I should instead do something like this :
> 
> ```
> module G = Group G
> module Glemmas = GroupLemmas G
> module H = Group H
> module Hlemmas = GroupLemmas H
> ```
> 
> But then it isn't obvious at all that I should use `G.invr` and `Glemmas.invId` and not the opposite. I think it just adds confusion to the end user experience.
> 
> Or maybe we should add all the notations and fields of the record type `Group` to the `GroupLemmas` module.
> But I think that too would reduce readability of the code. I think a comment explaining the choice to keep those properties there would make it all clearer and more practical to everyone.

Can't the `GroupLemmas` module just open the `Group` record publically? I think you would then get everything defined in Group when opening GroupLemmas.

Reply:
Thanks for the remark, I will change this right away.

Reply:
Great! Can you move as much of the other group theory there as well?

Reply:
I'm not sure I understand by "the other group theory" ?

Reply:
That seems like an outdated question from me